2단계 - 요구 사항 정리

jivebreaddev opened the pull request:
민영님 이번에도 잘 부탁드리겠습니다!

**구현사항**
- [x]  1. erd diagram 추가

- [x] 2. 구현사항들을 관찰해서 도메인에 대한 요구사항 정리

Jivebread committed:
chore(requirements): kitchenpos 요구사항 정리
- 1. erd diagram 추가
- 2. 구현사항들을 관찰해서 도메인에 대한 요구사항 정리

Gomding commented on README.md:
````diff
@@ -6,13 +6,127 @@
 docker compose -p kitchenpos up -d
 ```
+## DB 엔티티 다이어그램
+- 엔티티 내부의 필드는 추후에 갱신하겠습니다.
````
엔티티 다이어그램까지 그려주셨군요 👍 👍 👍 👍

Gomding commented on README.md:
```diff
@@ -6,13 +6,127 @@
 
+- 키친포스라는 식당을 운영하는데 필요한 POS기를 구현한다.
+- 메뉴 카테고리
+  - 액터  
```
액터 , 상태, 행동 3가지로 나눠서 표현해주셨군요 💯 💯 💯 

3가지로 표현해보니 명확해서 좋네요!

Gomding commented on README.md:
```diff
@@ -6,13 +6,127 @@
+    - 메뉴 카테고리는 이름을 가진다.
+  - 행동
+    - 사장님이 메뉴 카테고리를 생성할 수 있다.
+      - 생성 시 메뉴 카테고리의 이름은 빈값 혹은 NULL이 아니다.
```
`NULL` 이라는 것은 개발자 중심의 키워드라고 생각해요!
요구사항은 프로젝트 관계자 (개발자 + PM + 디자이너 + .. 등등등) 와 커뮤니케이션 할 때 사용한다고 생각하고 정리해주세요 :)
비개발자도 이해할 수 있는 키워드를 사용하는것이 중요합니다 🤗

```suggestion
        - 생성 시 메뉴 카테고리의 이름은 비어있을 수 없다
```

Gomding commented on README.md:
```diff
@@ -6,13 +6,127 @@
+  - 액터  
+    - 사장님
+  - 상태
+    - 메뉴 카테고리는 이름을 가진다.
```
요건 취향인 부분인데
상태가 가지는 요구사항이 있을거에요 :)

`메뉴 카테고리의 이름 은 1글자 이상이어야한다` 같은것이 이름의 요구사항 이겠네요!

Gomding commented on README.md:
```diff
@@ -6,13 +6,127 @@
 ## 요구 사항
 
+- 키친포스라는 식당을 운영하는데 필요한 POS기를 구현한다.
+- 메뉴 카테고리
```
메뉴 카테고리는 어떤 것인지 예시같은것도 활용해볼 수 있을 것 같아요 :)
모두가 이해할 수 있도록 이것저것 활용해보는것도 좋습니다 🤗

> ex) 치킨집이라면 메뉴판에 추천메뉴, 신메뉴, 한마리 메뉴, 두마리 메뉴, 순살 메뉴 로 그룹을 지어줄 수 있다.

Gomding commented on README.md:
```diff
@@ -6,13 +6,127 @@
+        - 사장님이 음식을 생성할 수 있다.
+            - 생성 시 메뉴 카테고리의 이름은 빈값은 가능하다.
+            - 비속어는 포함 될 수 없다.
+            - 가격은 0보다 작을 수 없다.
```
`가격은 음수가 될 수 없다` 라고도 표현할 수 있을것같아요!
사실 시창님이 표현해주신것도 문제는 없습니다 ㅎㅎ!

Gomding commented on README.md:
```diff
@@ -6,13 +6,127 @@
+        - 사장님이 상품들을 조회할 수 있다.
+        - 사장님이 상품 가격을 변경할 수 있다.
+          - 가격은 0 보다 작을 수 없다.
+          - 상품을 포함하는 메뉴의 가격이 올라가면 메뉴가 전시상태가 `미노출` 로 바뀐다.
```
해당 요구사항은 파악이 필요해보여요!
상품을 포함하는 `메뉴의 가격이 올라가면` 이 맞을까요? 🧐

jivebreaddev replied:
` 메뉴에 속한 상품들의 가격의 총합이 메뉴의 가격보다 높으면 메뉴가 전시상태가 `미노출` 로 바뀐다.` 로 변경하였습니다.

Gomding commented on README.md:
```diff
@@ -6,13 +6,127 @@
+      - 메뉴는 전시상태를 가진다.
+      - 메뉴는 상품들(1개 이상)을 가진다.
+    - 행동
+      - 사장님이 메뉴를 생성할 수 있다.
```
메뉴의 행동은 생성 외에 몇가지가 더 있었을거에요!
추가 부탁드려요 🤗

jivebreaddev replied:
1. 전시
2. 숨기기
3. 조회
를 추가했습니다.

Gomding commented on README.md:
```diff
@@ -6,13 +6,127 @@
+        - SERVED: 전달
+        - DELIVERING: 배달중
+        - DELIVERED: 배달완료
+        - COMPLETED: 주문완료
```
주문 상태 정리 👍

Gomding commented on README.md:
```diff
@@ -6,13 +6,127 @@
+        - DELIVERED: 배달완료
+        - COMPLETED: 주문완료
+      - 주문은 주문한 메뉴들을 가진다.
+      - 주문은 주문 종류를 가진다.
```
주문 종류 정리 👍

Gomding commented on README.md:
```diff
@@ -6,13 +6,127 @@
+        - EAT_IN 매장내
+    - 행동
+      - 사장님이 주문을 생성할 수 있다.
+        - 주문은 메뉴 1개 이상으로 구성되어야 한다.
```
`대기중` 은 언제 사용되는 상태일까요!

jivebreaddev replied:
`주문을 생성할 때, `대기중` 으로 주문 상태를 변경한다.`를 추가했습니다.

jivebreaddev replied:
`- 주문은 주문 상태를 가진다.
        - WAITING: 대기중 (주문 생성 시)
        - ACCEPTED: 접수 (주문 수락 시)
        - SERVED: 전달 (주문을 손님, 라이더에게 전달시)
        - DELIVERING: 배달중 (라이더에게 전달 완료)
        - DELIVERED: 배달완료 (라이더가 배달 완료)
        - COMPLETED: 주문완료 (주문완료 설정)`

Gomding commented on README.md:
```diff
@@ -6,13 +6,127 @@
+        - TAKEOUT - 포장
+        - EAT_IN 매장내
+    - 행동
+      - 사장님이 주문을 생성할 수 있다.
```
매장 식사(EAT_IN), 배달(DELIVERY) 이면 필수였던 정보가 있었을 거에요!
해당 정보도 정리되면 좋겠어요 😉

jivebreaddev replied:
추가적으로 정리 진행했습니다. 
`        - DELIVERY - 배달 (필수값: 주문 장소)
        - TAKEOUT - 포장 
        - EAT_IN 매장내 (필수값: 주문 테이블)
        `



      `- 사장님이 주문을 생성할 수 있다.
        - 주문은 메뉴 1개 이상으로 구성되어야 한다.
        - 주문 종류: 공통 (해당 조건은 공통으로 지켜져야 한다.)
          - 주문을 생성할 때, `대기중` 으로 주문 상태를 변경한다.
          - 메뉴 내의 주문상품들이 삭제된 상품이면 안된다.
          - 메뉴 내의 주문상품들이 삭제된 상품이면 안된다.
          - 메뉴가 생성된 상태이고 전시 되어있어야한다.
          - 메뉴의 가격 총합과 주문상품들의 가격 총합이 동일해야한다.
        - 주문 종류: 배달
          - 주문을 취소할 수 없다.
          - 주문 장소가 없다면 주문할 수 없다.
        - 주문 종류: 포장
          - 주문을 취소할 수 없다.
        - 주문 종류: 매장내
          - 주문 테이블이 없으면 주문 할 수 없다.
          - 주문 테이블이 있으면 테이블을 주문에 할당한다.
          - 주문을 취소할 수 있다.`

Gomding reviewed (changes requested):
안녕하세요 시창님!
이번 단계도 빠르게 진행해주셨군요 👍 
몇 가지 코멘트 남겼으니 확인 부탁드려요!

궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요 :)

jivebreaddev commented:
- [x] 1. NULL -> `비어있다` 같은 비개발자적 표현

- [x] 2. 상태가 가지는 요구사항도 표기해주기

- [x] 3. 메뉴 카테고리 같은것들은 예시도 적어주면 이해하기 편하다.

- [x] 4. 0 보다 작을 수 없다. == 음수가 될 수 없다.

- [x] 5. `대기중` 과 같은 상태에 대한 설명도 있으면 이해하기 편리하다.

- [x] 6. `매장 식사`, `배달`이면 필수인 정보들도 있으면 이해하기 편리하다.

- [x] 7. 메뉴 행동에서 누락된 부분 확인 필요하다.

- [x] 8. 메뉴의 가격이 올라가면 같은 요구사항이 틀린것으로 보이는 항목이 있다.

Jivebread committed:
chore(requirements): kitchenpos 요구사항 재정리
 1. NULL -> 비어있다 같은 비개발자적 표현
 2. 상태가 가지는 요구사항도 표기해주기
 3. 메뉴 카테고리 같은것들은 예시도 적어주면 이해하기 편하다.
 4. 0 보다 작을 수 없다. == 음수가 될 수 없다.
 5. 대기중 과 같은 상태에 대한 설명도 있으면 이해하기 편리하다.
 6. 매장 식사, 배달이면 필수인 정보들도 있으면 이해하기 편리하다.
 7. 메뉴 행동에서 누락된 부분 확인 필요하다.
 8. 메뉴의 가격이 올라가면 같은 요구사항이 틀린것으로 보이는 항목이 있다.

Gomding commented on README.md:
```diff
@@ -15,38 +15,38 @@ docker compose -p kitchenpos up -d
     - 사장님
   - 상태
-    - 메뉴 카테고리는 이름을 가진다.
+    - 메뉴 카테고리는 이름을 가진다. (ex: '추천메뉴', '신메뉴', '2명을 위한 메뉴')
```
예시를 전체에 추가해주셨군요 😃 
저는 이해할 수 없을것 같은 부분에 예시를 추가했었는데
전체적으로 예시를 추가해주니 이해하기 더 좋은 문서라는 생각이 드네요 👍 👍 👍 
덕분에 좋은 인사이트 얻어갑니다!

Gomding commented on README.md:
```diff
@@ -15,38 +15,38 @@ docker compose -p kitchenpos up -d
+            - 생성 시 메뉴 카테고리의 이름은 비어있을 수 있다.
             - 비속어는 포함 될 수 없다.
-            - 가격은 0보다 작을 수 없다.
+            - 가격은 양수이어야 한다.
```
👍 👍 👍 👍

Gomding commented on README.md:
```diff
@@ -6,13 +6,145 @@
+        - 상품은 가격을 가진다.
+    - 행동
+        - 사장님이 음식을 생성할 수 있다.
+            - 생성 시 메뉴 카테고리의 이름은 비어있을 수 있다.
```
상품이 메뉴 카테고리와 관계가 있었을까요! 🤔

jivebreaddev replied:
삭제 처리 했습니다.

제가 요구사항을 잘못읽었군요!

Gomding commented on README.md:
```diff
@@ -6,13 +6,145 @@
+        - 상품은 음식 이름을 가진다. (ex: '프라이드 치킨', '옥수수튀김', '회전회오리 튀김')
+        - 상품은 가격을 가진다.
+    - 행동
+        - 사장님이 음식을 생성할 수 있다.
```
여기서 음식은 상품을 의미하는 것일까요?
이미 아시겠지만 사용하는 단어를 통일시키는것이 중요해요 😃

jivebreaddev replied:
`상품은 상품 이름을 가진다. (ex: '프라이드 치킨', '옥수수튀김', '회전회오리 튀김')` 로 변경하였습니다.

Gomding commented on README.md:
```diff
@@ -56,32 +56,50 @@ docker compose -p kitchenpos up -d
-        - DELIVERING: 배달중
-        - DELIVERED: 배달완료
-        - COMPLETED: 주문완료
+        - WAITING: 대기중 (주문 생성 시)
```
상태의 시점(?)을 표현해주니 어떤 상태인지 더욱 명확해지네요 👍 👍 👍

Gomding commented on README.md:
```diff
@@ -6,13 +6,145 @@
+  - 액터
+    - 사장님
+  - 상태
+    - OCCUPIED: 사용중
```
주문 테이블도 몇가지 상태가 더 있었을 거에요 :)
행동에 이미 적혀있는것을 보니 깜빡하신것 같군요 😎 
추가만 해주시면 될 것 같아요!

Gomding replied:
+) 사용중 이라는 상태가 있다면 반대의 상태도 있을것 같아요!

어떤 단어로 소통하면 좋을지 정해주면 더욱 풍부한 대화가 가능하겠어요 👍

jivebreaddev replied:
`상태


   OCCUPIED: 사용중


   UNOCCUPIED: 미사용중


   주문 테이블은 손님의 수를 가진다.`
   
로 변경하였습니다.

Gomding commented on README.md:
```diff
@@ -6,13 +6,145 @@
+    - 사장님이 주문 테이블의 손님의 수를 변경한다.
+      - 주문 테이블의 변경하고자 하는 손님의 수가 음수라면 변경되지 않는다.
+      - 주문 테이블이 등록되어 있지 않으면 변경되지 않는다.
+    - 사장님이 주문 테이블의 상태를 사용가능하게 만든다.
```
`사용 가능한 테이블` 이란 어떤 상태를 의미할까요! 🤗

> 고객수가 몇명이며, 사용중 상태는 어떤 상태일까요?

jivebreaddev replied:
`손님의 수를 0으로 변경하여 주문 테이블의 상태를 `미사용중`으로 변경할 수 있다.` 
로 변경하였습니다.

Gomding commented on README.md:
```diff
@@ -6,13 +6,145 @@
+        - 주문 종류: 공통 (해당 조건은 공통으로 지켜져야 한다.)
+          - 주문을 생성할 때, `대기중` 으로 주문 상태를 변경한다.
+          - 메뉴 내의 주문상품들이 삭제된 상품이면 안된다.
+          - 메뉴 내의 주문상품들이 삭제된 상품이면 안된다.
```
중복된 내용이 있군요 😉

jivebreaddev replied:
삭제처리하였습니다.

Gomding commented on README.md:
```diff
@@ -6,13 +6,145 @@
+          - 메뉴 내의 주문상품들이 삭제된 상품이면 안된다.
+          - 메뉴 내의 주문상품들이 삭제된 상품이면 안된다.
+          - 메뉴가 생성된 상태이고 전시 되어있어야한다.
+          - 메뉴의 가격 총합과 주문상품들의 가격 총합이 동일해야한다.
```
주문 생성 시 아래와 같은 조건도 있었어요!

어떤 조건인지 파악이 필요합니다 :)

```java
        for (final OrderLineItem orderLineItemRequest : orderLineItemRequests) {
            final long quantity = orderLineItemRequest.getQuantity();
            if (type != OrderType.EAT_IN) {
                if (quantity < 0) {
                    throw new IllegalArgumentException();
                }
            }
```

jivebreaddev replied:
![image](https://github.com/next-step/ddd-legacy/assets/89639413/df4a2ce3-abc2-4ed0-98a4-a131abc09b82)

jivebreaddev replied:
`주문을 생성할 때, 주문상품들의 갯수를 감소시킬 수 없다.`
로 표기하였습니다.

매장 내 주문유형은 
`주문을 생성할 때, 주문상품들의 갯수를 감소시킬 수 있다.` 
다음으로 표기하였습니다.

Gomding replied:
헉 해당 요구사항은 아래와 같을것같아요

- 주문 종류가 매장 내(EAT_IN)가 아닐 때 주문한 메뉴들의 각 개수는 0개 이상이어야 한다

Gomding reviewed (changes requested):
안녕하세요 시창님!
꼼꼼하게 변경해주셨군요 :)
덕분에 저도 배울 수 있었습니다 👍 
추가로 코멘트 남겼으니 확인부탁드려요!

궁금한 점 있으면 편하게 DM 이나 코멘트 남겨주세요 😄

Jivebread committed:
chore(requirements): kitchenpos 요구사항 재정리
 1. 상태는 예시를 듦으로써 조금더 꼼꼼히 보여준다.
 2. 0보단 양수, 음수 사용
 3. `상품과 메뉴카테고리` 같은 오류 잡기
 4. 음식과 상품 같은 중복 명명 피하기
 5. 상태는 상태 시점을 정확히 표현하기
 6. BOOLEAN 값도 상태를 부여하여서 정확한 상태 지정하기
 7. 사용가능한 상태 -> 세세한 조건들 나열하기
 8. 중복제거
 9. 누락된 조건들 추가해서 작성하기

Gomding commented on README.md:
```diff
@@ -123,6 +121,8 @@ docker compose -p kitchenpos up -d
   - 상태
     - OCCUPIED: 사용중
+    - UNOCCUPIED: 미사용중
+    - 주문 테이블은 손님의 수를 가진다. 
```
주문 테이블에도 이름이 있으니 상태에 이름도 명시해줄 수 있겠네요 😃

Gomding commented on README.md:
```diff
@@ -131,7 +131,7 @@ docker compose -p kitchenpos up -d
       - 주문 테이블의 변경하고자 하는 손님의 수가 음수라면 변경되지 않는다.
       - 주문 테이블이 등록되어 있지 않으면 변경되지 않는다.
-    - 사장님이 주문 테이블의 상태를 사용가능하게 만든다.
+    - 사장님이 손님의 수를 초기화하여(0이 기본값) 주문 테이블의 상태를 `미사용중`으로 변경할 수 있다.
```
앗 이부분은 이전에 적어주신게 더 좋다고 생각해요
다음과 같이 명시하는것은 어떤지 제안드려봤었어요!

```
    - 사장님이 주문 테이블의 상태를 사용가능하게 만든다.
      - 사용 가능한 테이블이란 미사용 중 상태 + 손님이 0명인 상태이다.
```

jivebreaddev replied:
앗.. 제가 적은 포맷을 유지를 못했네요!! 감사합니다!

Gomding reviewed (approved):
안녕하세요 시창님!
이번 단계는 이만 merge 해도 될 것같아요!
추가로 남긴 코멘트 확인해주세요 🤗

다음 단계도 화이팅입니다~
